File: Version1.3/flaskServer/dash/graphing.py
from dash_html_components import Div, Span
from flask import render_template, request, g
from flask_login import current_user
from dash import Dash
from dash_core_components import (
    Graph, Slider, RangeSlider,
    DatePickerRange, Loading)
from dash_html_components import Div, Pre, Br, Button, Span
from dash_bootstrap_components import Tooltip
from dash.dependencies import (Input,
                               Output, State, ClientsideFunction)
from dash_extensions import Download
from dash_extensions.snippets import send_data_frame
from dash.exceptions import PreventUpdate
from pandas import read_csv, DataFrame
from ..extensions import collection, mongo, get_fs
from ..config import csv_label, rv_label
import julian
from datetime import date
import dash_daq as daq
from json import dumps
import plotly

url_base = '/data/'

rv_plot = None


def from_mjd(num):
    return str(julian.from_jd(float(num), fmt='mjd'))[0:23]


def to_mjd(num):
    return julian.to_jd(date(num), fmt='mjd')

def rv_plot(server):
    with server.app_context():

        entries = list(collection('radialvelocity').find())
        if entries == []:
            df_none = DataFrame(columns=[
                csv_label['mjd'], csv_label['filename'], csv_label['v']
            ])
            rv_plot.result = {}
            rv_plot.data = df_none
            rv_plot.public = df_none
            return df_none
        data = DataFrame(entries)
        data[csv_label['mjd']] = data[csv_label['mjd']].apply(from_mjd)
        
        rv_plot.data = data
        rv = data[data['PUBLIC'] == True]

        fig = plotly.graph_objs.Scatter(
            x= rv[csv_label['mjd']],
            y= rv[csv_label['v']],
            mode="markers",
            marker=plotly.graph_objs.scatter.Marker(
                opacity=0.6,
                colorscale="Viridis"
            )
        )
        x_axis = {
            'title': {
                'text': 'Date',
                'font': {
                    'size': 18,
                    'color': '#7f7f7f'
                }
            },
        }

        y_axis = {
            'title': {
                'text': 'Radial Velocity',
                'font': {
                    'size': 18,
                    'color': '#7f7f7f'
                }
            },
        }

        rv_plot.result = dumps({'data': [fig], 'layout':{'xaxis':x_axis, 'yaxis':y_axis}, 'config': {'responsive': True}}, cls=plotly.utils.PlotlyJSONEncoder)
        rv_plot.public = rv
        return rv

tool = Span(
    "floccinaucinihilipilification",
    id="tooltip-target",
    style={"textDecoration": "underline", "cursor": "pointer"},
)

def init_graphing(server):
    #fs = gridfs.GridFS(mongo.db)
    with server.app_context():
        one = get_fs('one')
        two = get_fs('two')
    
    rv = rv_plot(server)
    

    app = Dash(
        '__main__',
        server=server,
        url_base_pathname=url_base,
        assets_folder='static',
    )

    @server.before_request
    def update():
        
        if request.endpoint == url_base:
            app.index_string = render_template(
                'data_page.html'
            )
            rv = rv_plot.public
            curr_user = g.user['atype']
            if curr_user == 'admin' or curr_user == 'researcher':
                rv = rv_plot.data
            app.layout = Div([
                Tooltip(
                    "Set the Y axis to a linear or logarithmic scale.",
                    target="log-switch",
                ),
                Tooltip(
                    "Enables data to be separated by order on the spectrograph",
                    target="dim-switch",
                ),
                Tooltip(
                    "Each point represents a reading from the spectrograph "
                    "Clicking a point will load that spectrum's data",
                    target="rv-tool",
                ),
                Tooltip(
                    "The name of the file being displayed",
                    target="click-data",
                    placement="bottom"
                ),
                Tooltip(
                    "Selects which orders of the spectrum to render",
                    target="order-tool",
                ),
                Tooltip(
                    "Reduces the amount of points rendered to the screen",
                    target="res-tool",
                ),
                Tooltip(
                    "Each graph is interactive, and can be navigated with its toolbar",
                    target="spec-tool"
                ),
                Loading([
                    
                    Div(
                        id='rv-download-container'
                    ),
                    Button(
                        'Download Radial Velocities',
                        id='rv-download',
                        className='btn btn-primary btn-sm float-right',
                    ),
                    Download(
                        id='rv-download-data'
                    ),
                    DatePickerRange(
                        id='date-range',
                        className='pt-1 d-flex justify-content-end w-100',
                        min_date_allowed=date(2020, 8, 23),
                        max_date_allowed=date(2021, 9, 19),
                        initial_visible_month=date(2020, 8, 23),
                        end_date=date(2020, 9, 18)
                    ),
                    Span(['(i)'], id='rv-tool', className='point'),
                    Graph(
                        id='rv-plot',
                        className="",
                        config={
                            'modeBarButtonsToRemove': ['pan2d', 'lasso2d', 'autoscale']
                        }
                    ),
                    
                ], type="default", className='d-flex justify-content-end w-100'),
                Div(
                    id='rv-data',
                    children=rv[rv_label].to_json(),
                    className='d-none'
                ),
                Div([
                    Pre(id='click-data', className=' d-inline'),
                ]),
                
                
                Br(),
                Br(),
                Div([
                    Loading([
                        Div([
                            Div([
                                Button(
                                    'Download 1D',
                                    id='1d-spec-download',
                                    className='btn btn-primary btn-sm',
                                ),

                                Download(
                                    id='1d-spec-download-data'
                                ),
                            ], className='pr-2'),
                            Div([
                                Button(
                                    'Download 2D',
                                    id='2d-spec-download',
                                    className='btn btn-primary btn-sm',
                                ),

                                Download(
                                    id='2d-spec-download-data'
                                ),
                            ], className=""),
                        ], className="row justify-content-end d-none", id="spec-download-container"),
                        Div(
                            id='spec-data',
                            children=[],
                            className='d-none'
                        ),
                        Span(['(i)'], id='spec-tool', className='point'),
                        Graph(
                            id='spec-plot',
                            className="pt-0",
                            config={
                                'modeBarButtonsToRemove': ['pan2d', 'lasso2d', 'autoscale']
                            }
                        ),
                    ], type="default", ),


                    Div([
                        Span([
                            'Order Range\n'
                        ], id='order-tool', className='d-none'),
                        Span([
                            'Resolution\n'
                        ], id='res-tool', className='d-none')
                    ],
                    id='slide-label',
                    className='w-100 text-center'),


                    Div([
                        
                        Slider(
                            min=1,
                            max=200,
                            value=100,
                            marks={
                                1: {'label': '1:1', 'style': {'color': '#77b0b1'}},
                                20: {'label': '20:1'},
                                50: {'label': '50:1'},
                                100: {'label': '100:1'},
                                200: {'label': '200:1', 'style': {'color': '#f50'}}
                            },
                            id='resolution',
                            className='col d-none',
                        ),
                        RangeSlider(
                            id='spec-range',
                            className='col d-none',
                            min=0,
                            max=85,
                            step=1,
                            value=[40, 45],
                            marks={
                                0: {'label': '0'},
                                40: {'label': '40'},
                                45: {'label': '45'},
                                85: {'label': '85'}
                            }
                        ),
                    ], className='row'),
                ], id='spec-container'),
                Br(),
                Div([
                    Div([
                        daq.ToggleSwitch(
                            label='lin/log',
                            id='log-switch',
                            className='d-inline-block',
                            labelPosition='bottom'
                        ),
                    ], className='col text-center'),
                    Div([
                        daq.ToggleSwitch(
                            label='1D / 2D',
                            id='dim-switch',
                            className='d-inline-block',
                            labelPosition='bottom'
                        )
                    ], className='col text-center'),
                ], className="row")
            ])

    

    app.layout = Div([])

    app.clientside_callback(
        output=Output('click-data', 'children'),
        inputs=[Input('rv-plot', 'clickData'),
                State('rv-data', 'children')],
        clientside_function=ClientsideFunction(
            namespace='graphing',
            function_name='clickData'
        )
    )

    app.clientside_callback(
        output=Output('rv-plot', 'figure'),
        inputs=[Input('date-range', 'start_date'),
                Input('date-range', 'end_date'),
                State('rv-data', 'children')],
        clientside_function=ClientsideFunction(
            namespace='graphing',
            function_name='datefunc'
        )
    )

    app.clientside_callback(
        output=Output('spec-plot', 'figure'),
        inputs=[Input('resolution', 'value'),
                Input('spec-range', 'value'),
                Input('log-switch', 'value'),
                Input('spec-data', 'children')],
        clientside_function=ClientsideFunction(
            namespace='graphing',
            function_name='specfunc'
        )
    )


    @app.callback(Output('rv-download-data', 'data'),
                  Input('rv-download', 'n_clicks'))
    def rvDownload(n_clicks):
        if (n_clicks is not None) and (n_clicks > 0):
            return send_data_frame(rv.drop(['_id', 'PUBLIC'], axis=1).to_csv, filename="radial_velocities.csv")

    @app.callback(Output('1d-spec-download-data', 'data'),
                  Input('1d-spec-download', 'n_clicks'),
                  State('click-data', 'children'))
    def oneDownload(n_clicks, children):
        if (n_clicks is not None) and (n_clicks > 0) and (children is not None):
            searchDate = children.strip('Sun_')
            searchDate = searchDate.strip('.fits')
            try:
                specData = read_csv(one.find_one({'filename': {'$regex': '.*' + searchDate + '.*'}}))
                return send_data_frame(specData.to_csv, filename=searchDate+'.1d_spectrum.csv')
            except:
                entry = one.find_one({})
                specData = read_csv(entry)
                return send_data_frame(specData.to_csv, filename=entry['filename']+".1d_spectrum.csv")

    @app.callback(Output('2d-spec-download-data', 'data'),
                  Input('2d-spec-download', 'n_clicks'),
                  State('click-data', 'children'))
    def twoDownload(n_clicks, children):
        if (n_clicks is not None) and (n_clicks > 0) and (children is not None):
            searchDate = children.strip('Sun_')
            searchDate = searchDate.strip('.fits')
            try:
                specData = read_csv(two.find_one({'filename': {'$regex': '.*' + searchDate + '.*'}}))
                return send_data_frame(specData.to_csv, filename=searchDate + ".2d_spectrum.csv")
            except:
                entry = one.find_one({})
                specData = read_csv(entry)
                return send_data_frame(specData.to_csv, filename=entry['filename']+".2d_spectrum.csv")


    @app.callback(
        Output('spec-data', 'children'),
        Output('spec-download-container', 'className'),
        Output('spec-range', 'className'),
        Output('resolution', 'className'),
        Output('slide-label', 'children'),
        Input('click-data', 'children'),
        Input('dim-switch', 'value')
    )
    def getGraph(children, dim):

        if (children == None):
            raise PreventUpdate
        
        searchDate = children.strip('Sun_')
        searchDate = searchDate.strip('.fits')

        if (dim):
            slide = [
            Span([
                'Order Range\n'
            ], id='order-tool', className=''),
            Span([
                'Resolution\n'
            ], id='res-tool', className='d-none')
        ]
            regex = two.find_one({'filename': {'$regex': '.*' + searchDate + '.*'}})
            if regex:
                res = read_csv(regex).to_json()
                return res, 'row justify-content-end', 'col', 'col d-none', slide
            res = read_csv(two.find_one({})).to_json()
            return res, 'row justify-content-end', 'col', 'col d-none', slide

        slide = [
            Span([
                'Order Range\n'
            ], id='order-tool', className='d-none'),
            Span([
                'Resolution\n'
            ], id='res-tool', className='')
        ]
        regex = one.find_one({'filename': {'$regex': '.*' + searchDate + '.*'}})
        if regex:
            res = read_csv(regex).to_json()
            return res, 'row justify-content-end', 'col d-none', 'col', slide
        res = read_csv(one.find_one({})).to_json()
        return res, 'row justify-content-end', 'col d-none', 'col', slide

    return app

File: Version1.3/static/graphing.js

csv_label = {
    'accept': 'ACCEPT',
    'filename': 'FILENAME',
    'datetime': 'MJD',
    'velocity': 'V',
    'flux': 'FLUX',
    'wavelength': '# WAVE',
    'order': 'ORDER'
}
graph_label = {
    'rv_x': 'Recorded Date',
    'rv_y': 'Radial Velocity',
    'spec_x': 'Wavelength(Angstrom)',
    'spec_y': 'Flux',
}




window.dash_clientside = Object.assign({}, window.dash_clientside, {
    graphing: {
        clickData: function(clickData, table) {
            if(clickData === undefined) return;
            data = JSON.parse(table)
            return data[csv_label['filename']][clickData.points[0].pointIndex]
        },
        
        datefunc: function(start, end, table) {
            var xrange
            if(table === undefined) {
                throw "input undefined"
            }
            if(start === undefined) {
                xrange = {
                    
                    'autoscale': true}
            } else {
                xrange = {
                    
                    'range': [start, end]
                }
            }

            data = JSON.parse(table)
            console.log(start)
            return {
                'data':[{
                    'x': Object.values(data[csv_label['datetime']]),
                    'y': Object.values(data[csv_label['velocity']]),
                    'xaxis': 'x',
                    'yaxis': 'y',
                    'type': 'scattergl',
                    'mode': 'markers'
                }],
                'layout':{
                    'xaxis': {
                        'title': {
                            'text': graph_label['rv_x'],
                            'font': {
                                'size': 18,
                                'color': '#7f7f7f'
                            }
                        },
                        xrange,
                    },
                    'yaxis': {
                        'title': {
                            'text': graph_label['rv_y'],
                            'font': {
                                'size': 18,
                                'color': '#7f7f7f'
                            }
                        },
                        'autoscale': true
                    },
                    'type': 'scattergl',
                    "hovermode": "closest"
                }
            }

        },

        specfunc: function(res, range, log, table) {
            if(table === undefined) {
                throw "input undefined"
            }
            
            data = JSON.parse(table)

            let linlog = "linear"
            if(log) {
                linlog = "log"
            }
            let index = 0, x_order = [], y_order = [], data_structure = []

            if( data.hasOwnProperty(csv_label['order'])) {
                for(let order = range[0]; order < range[1]; order++) {
                    while(data[csv_label['order']][index] <= order) {
                        if(data[csv_label['order']][index] === order) {
                            x_order.push(data[csv_label['wavelength']][index])
                            y_order.push(data[csv_label['flux']][index])
                        }
                        index++
                    }

                    data_structure.push({
                        "line": { "color": ["#EE3124", "#F8971D", "#FFDD00", "#3DAE2B", "#00AEEF", "#002F87", "#A25EB5"][order%7], "dash": "solid" },
                        "mode": "lines",
                        "name": "<b>" + order + "</b>",
                        "showlegend": true,
                        "type": "scattergl",
                        "x": x_order,
                        "xaxis": "x",
                        "y": y_order,
                        "yaxis": "y"
                    })
                    x_order = []
                    y_order = []
                }
            } else {

                less = {
                    'wave':[],
                    'flux':[]
                }
                len =Object.keys(data[csv_label['wavelength']]).length
                for (i = 0; i < len; i=i+res) {
                    f=0
                    for (r=0; r < res; r+=1) {
                        f += data[csv_label['wavelength']][i+r]
                    }

                    less['wave'].push(f/res);
                    less['flux'].push(data[csv_label['flux']][i]);
                }
                data_structure = [{
                    'x': Object.values(less['wave']),
                    'y': Object.values(less['flux']),
                    'xaxis': 'x',
                    'yaxis': 'y',
                    'type': 'linegl',
                }]
            }
            
            
            return {
                'data': data_structure,
                'layout':{
                    'yaxis': {
                        'title': {
                            'text': graph_label['spec_y'],
                            'font': {
                                'size': 18,
                                'color': '#7f7f7f'
                            }
                        },
                        'type': linlog,
                        'autoscale': true},
                    'xaxis': {
                        'title': {
                            'text': graph_label['spec_x'],
                            'font': {
                                'size': 18,
                                'color': '#7f7f7f'
                            }
                        },
                        'autoscale': true
                    },
                    'type': 'linegl',
                    "hovermode": "closest",
                    
                }
            }
        },
    }
});
